qualtran/surface_code/t_factory_utils.py

```diff
@@ -0,0 +1,91 @@
+#  Copyright 2024 Google LLC
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#      https://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+
+from typing import Sequence
+
+import cirq
+import numpy as np
+
+
+def _rotation_unitary(pauli_string: str, theta: float):
+    num_qubits = len(pauli_string)
+    u = cirq.unitary(cirq.DensePauliString(pauli_string))
+    return np.cos(theta) * np.eye(1 << num_qubits) + 1j * np.sin(theta) * u
+
+
+class PauliRotationChannel(cirq.Gate):
+    r"""A channel that applies a pi/8 pauli rotation with possible overshooting to 5pi/8, -pi/8, and 3pi/8.
+
+    The channel is defined as
+
+    $$
+        \sum_{k \in \{1, 5, -1, 3\}} p_k e^{-i \frac{\pi}{8} k P} \rho e^{i \frac{\pi}{8} k P}
+    $$
+    """
+
+    def __init__(self, pauli_string, p1, p2, p3):
```

Comment:
type annotations and document these arguments. E.g.

```
Args:
  pauli_string: string of 'x' 'y' 'z' 'i' ...
  p1: the probability of doing a 5pi/8 rotation
```

Reply:
done